lib/src/revset.rs

```diff
@@ -1648,19 +1648,41 @@ impl SymbolResolver for FailingSymbolResolver {
     }
 }
 
+pub type PrefixResolver<'a, T> = Box<dyn Fn(&HexPrefix) -> PrefixResolution<T> + 'a>;
```

Comment:
This can be of `Fn(&dyn Repo, &HexPrefix) -> ...` type. Then `&dyn Repo` reference can be removed from `IdPrefixContext` and `unsafe {}` cast in CLI will go away.

A downside is that the caller needs to guarantee the `repo` reference passed to `IdPrefixContext` is unchanged, but I don't think that's worse than the current design. We'll anyway need to ensure that `IdPrefixContext` is initialized with the same `repo` as the one `DefaultSymbolResolver` holds.

Reply:
FYI, I'm looking into this now.

Reply:
I made that change in a commit on top. Let me know if you prefer that I absorb it into the other commits.

Reply:
LGTM, thanks!